### ISSUE TITLE
Persistent Class TypeError

### ISSUE THREAD

**Issue by arnicas** [msg-systems/coreferee#8]
With most documents of longer than a few sentences (news articles), I am getting a recurrent error:

````
text = """
France retains its centuries-long status as a global centre of art, science, and philosophy, says Ljubomir Geric. He also notes it
hosts the world's fifth-largest number of UNESCO World Heritage Sites and is the leading tourist destination, 
receiving over 89 million foreign visitors in 2018. France is a developed country with the world's 
seventh-largest economy by nominal GDP, and the ninth-largest by PPP. In terms of aggregate household wealth, 
it ranks fourth in the world. France performs well in international rankings of education, health care, 
life expectancy, and human development. It remains a great power in global affairs, being one of the five 
permanent members of the United Nations Security Council (UNSC) and an official nuclear-weapon state. France is a 
founding and leading member of the European Union (EU) and the Eurozone, and a member of the Group of 7, North 
Atlantic Treaty Organization (NATO), Organisation for Economic Co-operation and Development (OECD), and the 
World Trade Organization (TWO).
"""
doc = nlp(text)

Unexpected error annotating document, skipping ....
<class 'TypeError'>
Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
  File "/opt/conda/envs/nlp/lib/python3.8/site-packages/coreferee/manager.py", line 110, in __call__
    self.annotator.annotate(doc)
  File "/opt/conda/envs/nlp/lib/python3.8/site-packages/coreferee/annotation.py", line 270, in annotate
    self.tendencies_analyzer.score(doc, self.keras_ensemble)
  File "/opt/conda/envs/nlp/lib/python3.8/site-packages/coreferee/tendencies.py", line 390, in score
    keras_inputs, scoring_necessary = self.prepare_keras_data([doc])
  File "/opt/conda/envs/nlp/lib/python3.8/site-packages/coreferee/tendencies.py", line 326, in prepare_keras_data
    self.get_vectors(potential_referred, doc)
  File "/opt/conda/envs/nlp/lib/python3.8/site-packages/coreferee/tendencies.py", line 263, in get_vectors
    this_object_vector = np.mean( np.array([t.vector for t in tokens]), axis=0)
  File "cupy/core/core.pyx", line 1188, in cupy.core.core.ndarray.__array__
````

My numpy is 1.19.5.  tensorflow 2.4.2.  I wonder if this is what your issue about versions being too permissive was about?  I'll try to replicate those more restricted installs.

**Comment by richardpaulhudson** [comment]
Hi Lynn, thanks for your message. Tensorflow 2.4.2 was indeed only released a few days ago and seemed like a likely culprit, so I upgraded to it, but unfortunately I was unable to reproduce the problem:

`>>> import coreferee, spacy
>>> nlp = spacy.load('en_core_web_trf')
>>> nlp.add_pipe('coreferee')
>>> doc = nlp("""
... France retains its centuries-long status as a global centre of art, science, and philosophy, says Ljubomir Geric. He also notes it
... hosts the world's fifth-largest number of UNESCO World Heritage Sites and is the leading tourist destination,
... receiving over 89 million foreign visitors in 2018. France is a developed country with the world's
... seventh-largest economy by nominal GDP, and the ninth-largest by PPP. In terms of aggregate household wealth,
... it ranks fourth in the world. France performs well in international rankings of education, health care,
... life expectancy, and human development. It remains a great power in global affairs, being one of the five
... permanent members of the United Nations Security Council (UNSC) and an official nuclear-weapon state. France is a
... founding and leading member of the European Union (EU) and the Eurozone, and a member of the Group of 7, North
... Atlantic Treaty Organization (NATO), Organisation for Economic Co-operation and Development (OECD), and the
... World Trade Organization (TWO).
... """)
>>> doc._.coref_chains.print()
0: France(1), its(3), it(27), France(59), France(100), It(120), France(154)
1: Geric(22), He(24)
2: world(66), world(98)
>>> import numpy
>>> print(numpy.__version__)
1.19.5
>>> import tensorflow
>>> print(tensorflow.__version__)
2.4.2
>>>`

If you or anyone else can get to the bottom of the problem, I would be most grateful!

Best wishes

Richard

**Comment by arnicas** [comment]
Hmm.  I replicated it with my env stuff pinned to the versions you used.  So it must be something deeper in my env. If I have time to look further, I will!

**Comment by arnicas** [comment]
Ah, can you tell me what version of cupy you are using?

**Comment by richardpaulhudson** [comment]
I'm not using cupy - I'm working on Windows with Cuda v11.3.109. There is a nightly test build that takes place on a Linux cloud server which is also working fine, but the cloud server does not have any GPUs. So the problem seems to be specific to certain GPU architectures / drivers.

**Comment by arnicas** [comment]
Ok, well, the internet is full of people asking about this: it's definitely a fact that cupy won't do numpy arrays on the GPU without some special handling:  See https://docs.cupy.dev/en/stable/user_guide/basic.html.  I doubt I'll be the only one hitting this one...  I can try to modify your code to get it to run locally if I have time :)

**Comment by richardpaulhudson** [comment]
It would be great if you or anyone else could add Cupy compatibility as a feature. :-)

**Comment by richardpaulhudson** [comment]
I will be releasing a new version of Coreferee within the next few weeks that uses Thinc instead of TensorFlow and which is fully tested to run on GPU as well as CPU.

**Comment by richardpaulhudson** [comment]
This should all be resolved now with v1.2.0.
